month 8/step 2

frrenzy opened the pull request:
- feat: add ingredients store
- refactor: move detail modal logic to store
- refactor: move order logic to store
- refactor: move constructor logic to store
- refactor: move tabs logic to store
- feat: add DnD capability to construct burger
- feat: add constructor DnD reorder
- feat: add Loading spinner for requests
- feat: add placeholder text for empty constructor

frrenzy committed:
feat: add ingredients store

frrenzy committed:
refactor: move detail modal logic to store

frrenzy committed:
refactor: move order logic to store

frrenzy committed:
refactor: move constructor logic to store

frrenzy committed:
refactor: move tabs logic to store

frrenzy committed:
feat: add DnD capability to construct burger

frrenzy committed:
feat: add constructor DnD reorder

frrenzy committed:
feat: add Loading spinner for requests

frrenzy committed:
feat: add placeholder text for empty constructor

frrenzy committed:
fix: incorrect use of splice

gennady-bars commented on package.json:
```diff
@@ -7,7 +7,14 @@
     "@ya.praktikum/react-developer-burger-ui-components": "^1.13.1",
     "prop-types": "^15.8.1",
     "react": "^18.2.0",
+    "react-dnd": "^16.0.1",
```
**Можно лучше**  (какие-то уже у Вас выполнены, возможно)

* Лучше задавать **фиксированные версии** зависимостям. Если удалить `package-lock.json` и попытаться повторно установить зависимости с помощью команды `npm install`, то **npm** установит последние версии зависимостей, что может привести к ошибкам.
* В `devDependencies` указывается то, что необходимо для сборки и разработки приложения, а в `dependencies` нужно указывать то, что необходимо непосредственно для работы приложения (express, uuid). Все линтеры, тесты и тд нужно только в `devDependencies` указывать, чтобы зря не раздувать кол-во пакетов в собранном для деплоя приложении.
* В проекте в некоторых файлах есть ошибка `EOF`. Для `git` важно наличие **пустой строки** в конце файла. Подробнее тут: https://stackoverflow.com/questions/5813311/whats-the-significance-of-the-no-newline-at-end-of-file-log
* хорошей практикой является подключение `normalize.css`, чтобы **унифицировать** стилизацию под разные браузеры. Подробнее тут https://htmlacademy.ru/blog/html/about-normalize-css . И эти стили нужно получать **первее**всех других, чтобы переопределить изначально стили по умолчанию.
* Лучше не обращаться к **нативному** тегу, а задать для него **отдельный** css-класс.
При изменении **тэгов** в шаблоне в таком случае код будет продолжать работать, если в стили прописаны через имена классов
[Хорошая статья](https://frontstuff.io/you-need-to-stop-targeting-tags-in-css) не эту тему.
* все **цвета** лучше вынести в `css`-переменные и использовать их, чтобы можно было мгновенно понимать, что это за цвет, и можно было бы подкорректировать его в одном месте  и не искать дублирования во всем коде

* Чтобы сделать сайт более доступным для пользователя, нужно стараться придерживаться **семантической верстки**.
Несколько полезных статей на эту тему:
•	[Что такое семантическая вёрстка и зачем она нужна](https://htmlacademy.ru/blog/html/semantics)
•	[Семантика для циников](https://www.youtube.com/watch?v=B5pomLr059U)

frrenzy replied:
+

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -1,119 +1,121 @@
+          type: ADD_TO_ORDER,
+          ingredient: {
+            ...ingredient,
+            uuid: uuidv4(),
```
Правильное место для создания `uuid`

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -1,119 +1,121 @@
-              isLocked={true}
+          {cart.map(({ name, price, image, _id, uuid }, idx) => (
+            <ConstructorTile
+              key={uuid}
```
Отлично, что указываете **уникальный** `id` для ингредиентов в заказе. Это гарантирует 100% логику `DnD`

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -1,13 +1,22 @@
-const Modal = ({ closeModal, children }) => {
+const Modal = ({ children }) => {
   const modalRoot = document.querySelector('#modal')
+  const dispatch = useDispatch()
```
Внутри компонента `Modal` и `ModalOverlay` не стоит использовать `dispatch` для закрытия попапов, так как для каждого отдельного окна обычно разная логика закрытия (и в следующих спринтах как раз это и будет). Вам нужно передавать функцию закрытия `onClose` в `props` модальному окну, как это было раньше

frrenzy replied:
+

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -1,13 +1,22 @@
+
+  const closeModal = useCallback(() => {
+    dispatch({ type: RESET_DETAIL })
+    dispatch({ type: CLOSE_ORDER_MODAL })
```
`Modal` является **универсальным** компонентом для любых модальных окон. Он не должен знать внутри себя какие-то **конкретные** данные (экшены) для конкретных попапов.

frrenzy replied:
+

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная


* Отлично, что указываете **уникальный** `id` для ингредиентов в заказе. Это гарантирует 100% логику `DnD`
* Отличный лоадер во время запроса
* Хорошая логика `DnD` и редакса


но есть некоторые недочеты: 


* если нет еще заказа и нажать на кнопку отправки на сервер, то приложение ломается. Скрин https://disk.yandex.ru/i/wf6ky2yRteIe-A 
* Внутри компонента `Modal` и `ModalOverlay` не стоит использовать `dispatch` для закрытия попапов, так как для каждого отдельного окна обычно разная логика закрытия (и в следующих спринтах как раз это и будет). Вам нужно передавать функцию закрытия `onClose` в `props` модальному окну, как это было раньше
* `Modal` является **универсальным** компонентом для любых модальных окон. Он не должен знать внутри себя какие-то **конкретные** данные (экшены) для конкретных попапов.



**Можно лучше**

*  желательно **очищать конструктор** после успешного получения номера заказа с сервера в блоке `then`, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты
* лучше не давать нажать на кнопку `Оформить заказ`, пока пользователь не добавил булку в заказ.
* Чтобы **не дублировать** типы экшенов и сами объекты с `type`, принято выносить их в отдельные функции, называемые **`action creator`** (функция, которая возвращает объект с `type`). Это позволяет сделать **понятное название** для функции, которое будет говорить, что делает она, и избавит от повтора объекта с `type` в коде. Подробнее тут: [https://rajdee.gitbooks.io/redux-in-russian/content/docs/basics/Actions.html#генераторы-экшенов-action-creators](https://rajdee.gitbooks.io/redux-in-russian/content/docs/basics/Actions.html#%D0%B3%D0%B5%D0%BD%D0%B5%D1%80%D0%B0%D1%82%D0%BE%D1%80%D1%8B-%D1%8D%D0%BA%D1%88%D0%B5%D0%BD%D0%BE%D0%B2-action-creators)



Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

frrenzy commented:
TODO: add counter reset

frrenzy committed:
fix: review #1

1) Disable order button if cart is empty
2) Move order request to button onClick from modal onMount
3) Add cart reset on successful order request

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
